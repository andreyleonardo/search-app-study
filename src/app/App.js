import React, { Component } from 'react';
import { isEmpty } from 'lodash';
import PropTypes from 'prop-types';
import SearchBar from './search-bar/SearchBar';
import SearchResult from './search-result/SearchResult';
import stubData from '../data.json';
import { backgroundImage } from '../pictures';
import './App.scss';

const MIN_QUERY_SIZE = 3;

class App extends Component {
  constructor(props) {
    super(props);

    this.state = {
      query: '',
      results: []
    };

    this.onChange = this.onChange.bind(this);
    this.handleSubmit = this.handleSubmit.bind(this);
  }

  onChange({ target: { value } }) {
    const query = value;

    this.setState({ query });
    this.handleSubmit({ query });
  }

  handleSubmit(param) {
    const { query } = param || this.state;
    let results = [];

    if (this.isQueryValid(query)) {
      results = this.findTours(query);
    }

    this.setState({ results });
  }

  isQueryValid(query) {
    return query.length >= MIN_QUERY_SIZE;
  }

  findTours(query) {
    const { data: { tours } } = this.props;
    return tours.filter(tour => tour.title.toLowerCase().match(query.toLowerCase()));
  }

  render() {
    const { results, query } = this.state;

    return (
      <div className="app">
        <header className="app__header">
          <h1>Experience Finder</h1>
        </header>
        <main className="app__main row center-xs">
          <picture className="app__background">
            <img className="app__background_image" src={backgroundImage} alt="main section" />
          </picture>
          <section className="app__search_bar col-xs-12">
            <div className="row center-xs">
              <div className="col-xs-12 col-sm-7">
                <SearchBar onChange={this.onChange} handleSubmit={this.handleSubmit} />
              </div>
            </div>
          </section>
          {this.isQueryValid(query) && isEmpty(results) ? (
            <section className="app__no_results col-xs-12">
              <p>No results found! =(</p>
            </section>
          ) : (
            <section className="app__search_results col-xs-12">
              { results.map((result, index) => (
                <SearchResult
                  key={result.title}
                  result={result}
                  index={index}
                  tabIndex={index + 1}
                />
              )) }
            </section>
          )}
        </main>
      </div>
    );
  }
}

App.propTypes = {
  data: PropTypes.shape({
    tours: PropTypes.arrayOf(
      PropTypes.shape({
        title: PropTypes.string
      })
    )
  })
};

App.defaultProps = {
  data: stubData
};

export default App;
