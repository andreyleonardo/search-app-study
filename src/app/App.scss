.app {
  text-align: center;
  height: 100vh;
  font-size: calc(10px + 2.5vmin);
  color: white;

  &__header {
    justify-content: center;
  }
}
