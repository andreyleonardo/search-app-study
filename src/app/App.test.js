import React from 'react';
import App from './App';

describe('App', () => {
  it('renders without crashing', () => {
    const app = shallow(<App />);
    expect(app).toMatchSnapshot();
  });
});
