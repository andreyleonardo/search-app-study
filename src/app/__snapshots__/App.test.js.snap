// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App renders without crashing 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <App
    data={
      Object {
        "tours": Array [
          Object {
            "currency": "$",
            "description": "Lorem ipsum dolor sit amet, consectetur adipisicing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.",
            "isSpecialOffer": false,
            "price": "14",
            "rating": "4.8",
            "title": "German Tour: Parliament Quarter &amp; Reichstag glass dome",
          },
          Object {
            "currency": "$",
            "description": "Lorem ipsum dolor sit amet, consectetur adipisicing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.",
            "isSpecialOffer": false,
            "price": "21",
            "rating": "4.8",
            "title": "Skip the Line: Pergamon Museum Berlin Tickets",
          },
          Object {
            "currency": "$",
            "description": "Lorem ipsum dolor sit amet, consectetur adipisicing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.",
            "isSpecialOffer": false,
            "price": "10",
            "rating": "4.6",
            "title": "Berlin WelcomeCard: Transport, Discounts &amp; Guide Book",
          },
          Object {
            "currency": "$",
            "description": "Lorem ipsum dolor sit amet, consectetur adipisicing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.",
            "isSpecialOffer": false,
            "price": "143",
            "rating": "4.6",
            "title": "Skip the Line: Berlin TV Tower Ticket",
          },
          Object {
            "currency": "$",
            "description": "Lorem ipsum dolor sit amet, consectetur adipisicing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.",
            "isSpecialOffer": false,
            "price": "14",
            "rating": "4.5",
            "title": "1-Hour City Tour by Boat with Guaranteed Seating",
          },
          Object {
            "currency": "$",
            "description": "Lorem ipsum dolor sit amet, consectetur adipisicing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.",
            "isSpecialOffer": false,
            "price": "210",
            "rating": "4.3",
            "title": "Berlin Hop-on Hop-off Bus Tour: 1- or 2-Day Ticket",
          },
          Object {
            "currency": "$",
            "description": "Lorem ipsum dolor sit amet, consectetur adipisicing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.",
            "isSpecialOffer": false,
            "price": "59",
            "rating": "4.8",
            "title": "German Tour: Reichstag with Plenary Chamber &amp; Cuppola",
          },
          Object {
            "currency": "$",
            "description": "Lorem ipsum dolor sit amet, consectetur adipisicing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.",
            "isSpecialOffer": true,
            "price": "41",
            "rating": "4.5",
            "title": "Berlin: 2.5-Hour Boat Tour Along the River Spree",
          },
          Object {
            "currency": "$",
            "description": "Lorem ipsum dolor sit amet, consectetur adipisicing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.",
            "isSpecialOffer": false,
            "price": "14",
            "rating": "4.5",
            "title": "Museum Pass Berlin: 3-Day Entry to Over 40 Top Museums",
          },
          Object {
            "currency": "$",
            "description": "Lorem ipsum dolor sit amet, consectetur adipisicing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.",
            "isSpecialOffer": true,
            "price": "50",
            "rating": "4.6",
            "title": "Reichstag: Rooftop Coffee Break at KÃ¤fer",
          },
          Object {
            "currency": "$",
            "description": "Lorem ipsum dolor sit amet, consectetur adipisicing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.",
            "isSpecialOffer": true,
            "price": "46",
            "rating": "4.9",
            "title": "Skip the Line: Neues Museum Berlin Tickets",
          },
          Object {
            "currency": "$",
            "description": "Lorem ipsum dolor sit amet, consectetur adipisicing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.",
            "isSpecialOffer": false,
            "price": "87",
            "rating": "4.8",
            "title": "German Tour: Reichstag with dome Chamber &amp;",
          },
          Object {
            "currency": "$",
            "description": "Lorem ipsum dolor sit amet, consectetur adipisicing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.",
            "isSpecialOffer": false,
            "price": "8",
            "rating": "4.3",
            "title": "Berlin Hop-on Hop-off Bus Tour with Live Commentary",
          },
          Object {
            "currency": "$",
            "description": "Lorem ipsum dolor sit amet, consectetur adipisicing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.",
            "isSpecialOffer": false,
            "price": "140",
            "rating": "4.5",
            "title": "Skip the Line: TV Tower Early Bird Tickets",
          },
        ],
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "children": <div
        className="app"
      >
        <Header
          onChangeLanguage={[Function]}
          selectedLanguage="en-US"
        />
        <main
          className="app__main row center-xs"
        >
          <picture
            className="app__background"
          >
            <img
              alt="main section"
              className="app__background_image"
              src="app_background_image.jpeg"
            />
          </picture>
          <SearchBar
            handleSubmit={[Function]}
            name="search_bar_input"
            onChange={[Function]}
          />
          <section
            className="app__search_results col-xs-12"
          />
        </main>
      </div>,
      "locale": "en-US",
      "messages": Object {
        "header.title": "Experience Finder",
        "no_results.message": "No results found! =(",
        "result_card_details.pricing": "{currency} {price}",
        "result_card_details.rating": "{rating}/5",
        "result_card_details.special_offer": "Special Offer",
        "search_bar.label": "Let's search for a great experience",
      },
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <Header
            onChangeLanguage={[Function]}
            selectedLanguage="en-US"
          />,
          <main
            className="app__main row center-xs"
          >
            <picture
              className="app__background"
            >
              <img
                alt="main section"
                className="app__background_image"
                src="app_background_image.jpeg"
              />
            </picture>
            <SearchBar
              handleSubmit={[Function]}
              name="search_bar_input"
              onChange={[Function]}
            />
            <section
              className="app__search_results col-xs-12"
            />
          </main>,
        ],
        "className": "app",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "onChangeLanguage": [Function],
            "selectedLanguage": "en-US",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <picture
                className="app__background"
              >
                <img
                  alt="main section"
                  className="app__background_image"
                  src="app_background_image.jpeg"
                />
              </picture>,
              <SearchBar
                handleSubmit={[Function]}
                name="search_bar_input"
                onChange={[Function]}
              />,
              <section
                className="app__search_results col-xs-12"
              />,
            ],
            "className": "app__main row center-xs",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": <img
                  alt="main section"
                  className="app__background_image"
                  src="app_background_image.jpeg"
                />,
                "className": "app__background",
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "alt": "main section",
                  "className": "app__background_image",
                  "src": "app_background_image.jpeg",
                },
                "ref": null,
                "rendered": null,
                "type": "img",
              },
              "type": "picture",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "handleSubmit": [Function],
                "name": "search_bar_input",
                "onChange": [Function],
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": Array [],
                "className": "app__search_results col-xs-12",
              },
              "ref": null,
              "rendered": Array [],
              "type": "section",
            },
          ],
          "type": "main",
        },
      ],
      "type": "div",
    },
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "children": <div
          className="app"
        >
          <Header
            onChangeLanguage={[Function]}
            selectedLanguage="en-US"
          />
          <main
            className="app__main row center-xs"
          >
            <picture
              className="app__background"
            >
              <img
                alt="main section"
                className="app__background_image"
                src="app_background_image.jpeg"
              />
            </picture>
            <SearchBar
              handleSubmit={[Function]}
              name="search_bar_input"
              onChange={[Function]}
            />
            <section
              className="app__search_results col-xs-12"
            />
          </main>
        </div>,
        "locale": "en-US",
        "messages": Object {
          "header.title": "Experience Finder",
          "no_results.message": "No results found! =(",
          "result_card_details.pricing": "{currency} {price}",
          "result_card_details.rating": "{rating}/5",
          "result_card_details.special_offer": "Special Offer",
          "search_bar.label": "Let's search for a great experience",
        },
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <Header
              onChangeLanguage={[Function]}
              selectedLanguage="en-US"
            />,
            <main
              className="app__main row center-xs"
            >
              <picture
                className="app__background"
              >
                <img
                  alt="main section"
                  className="app__background_image"
                  src="app_background_image.jpeg"
                />
              </picture>
              <SearchBar
                handleSubmit={[Function]}
                name="search_bar_input"
                onChange={[Function]}
              />
              <section
                className="app__search_results col-xs-12"
              />
            </main>,
          ],
          "className": "app",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "onChangeLanguage": [Function],
              "selectedLanguage": "en-US",
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                <picture
                  className="app__background"
                >
                  <img
                    alt="main section"
                    className="app__background_image"
                    src="app_background_image.jpeg"
                  />
                </picture>,
                <SearchBar
                  handleSubmit={[Function]}
                  name="search_bar_input"
                  onChange={[Function]}
                />,
                <section
                  className="app__search_results col-xs-12"
                />,
              ],
              "className": "app__main row center-xs",
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": <img
                    alt="main section"
                    className="app__background_image"
                    src="app_background_image.jpeg"
                  />,
                  "className": "app__background",
                },
                "ref": null,
                "rendered": Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "alt": "main section",
                    "className": "app__background_image",
                    "src": "app_background_image.jpeg",
                  },
                  "ref": null,
                  "rendered": null,
                  "type": "img",
                },
                "type": "picture",
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "function",
                "props": Object {
                  "handleSubmit": [Function],
                  "name": "search_bar_input",
                  "onChange": [Function],
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": Array [],
                  "className": "app__search_results col-xs-12",
                },
                "ref": null,
                "rendered": Array [],
                "type": "section",
              },
            ],
            "type": "main",
          },
        ],
        "type": "div",
      },
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
